woocommerce/woocommerce-ios
# Issue/5159 show product category filter list

Part of https://github.com/woocommerce/woocommerce-ios/issues/5159

## Description

In the context of https://github.com/woocommerce/woocommerce-ios/issues/5159, we implement in this PR the `FilterProductCategoryListViewController` that shows the categories plus the "Any" filter row. Furthermore, we enhance the `productCategories` case of `FilterListValueSelectorConfig` with the `siteID` `Int64` necessary to fetch the relevant categories. Likewise, we pass that `siteID` to the `FilterProductListViewModel` from the `ProductsListViewController`, so it can be passed along. While the selection UI is implemented here, the selected product category is not passed to the filter yet, so products are not filtered by category yet. (will be implemented in the next PR)

## Challenges

In https://github.com/woocommerce/woocommerce-ios/pull/5230 we extracted the `ProductCategoryListViewController` from `EditProductCategoryListViewController`, so it can be reused in other cases, like in this one. There are however few features that are only specific to the filter product category list case:
- We should show an extra row on top with the title "Any"
- Unlike the case of `EditProductCategoryListViewController` where none or multiple categories can be selected, in this case **only and always one** should be selected (or Any)

To implement this business logic, we extend the functionality of `ProductCategoryListViewModel` with two protocols, 
`ProductCategoryListViewModelEnrichingDataSource`, that can modify the UI, and `ProductCategoryListViewModelDelegate`, that reacts to the selection event. The relevant logic for our case is encapsulated in `FilterProductCategoryListViewModel`, that conforms to these two protocols. It adds an extra row on top of the product category list (Any), and reacts to the selection event by reseting the categories so only one is selected, and updating the Any row selection state.  As you can see, nothing changes for the `EditProductCategoryListViewModel`, thus keeping its behaviour exactly as it was.

## Changes
- Pass that `siteID` to the `FilterProductListViewModel` from the `ProductsListViewController`
- Enhance the `productCategories` case of `FilterListValueSelectorConfig` with the `siteID` `Int64`
- Use it to retrieve product categories linked to that `siteID`
- Create `FilterProductCategoryListViewController` to show the product categories to be filtered
- Create `FilterProductCategoryListViewModel` to handle its logic
- Extend `ProductCategoryListViewModel` with `ProductCategoryListViewModelEnrichingDataSource` and `ProductCategoryListViewModelDelegate`
- Make `FilterProductCategoryListViewModel` conform to these protocols to implement its specific business logic
- Make `categoryID` in `ProductCategoryCellViewModel` optional so we can handle the "Any" cell case, that is not linked to any particular product category
- Refactor `ProductCategoryListViewController` initializer to accept the view model as parameter instead of multiple ones that are only used in the view model
- Move `MockProductCategoryStoresManager` to its own file, make it public
- Add Unit tests, TODOs

## Test
- Open the app
- Go to Products tab
- Open Filter
- Open Product Category:
  - Loading ghost effect should be visible
  - Product Categories are shown
  - Any row is shown on top
  - Only and always one product category (or the any row) can be selected (same as other filters)
  - UI looks good for landscape and portrait

<img src="https://user-images.githubusercontent.com/1864060/138704731-97ad31f7-6869-4c05-bc26-f541e036946d.png" width="320" height="693">
<img src="https://user-images.githubusercontent.com/1864060/138704951-680260b7-3df1-4470-8e73-990397adf349.png" width="693" height="320">

## Next Steps
- Pass the selected product category along and filter the products according to it
- Remove the feature flag

Update release notes:
